### PR TITLE
Add niid-Bench baseline

### PR DESCRIPTION
**Pull request by aashishkolluri**
### Description
Originally, the task has been to reproduce the results from SCAFFOLD paper (Karimireddy el al. 2019). The results in that paper are difficult to reproduce due to missing details on the experimental setup. However, SCAFFOLD has been benchmarked in another work (https://arxiv.org/pdf/2102.02079.pdf) along with other FL protocols FedAvg, FedProx, and FedNova. Therefore, in the best interests of demonstrating SCAFFOLD on Flower, I have implemented the other FL protocols too and reproduced results from this work instead.
 

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:80]:
```suggestion
# Set python version
pyenv local 3.10.6
# Tell poetry to use python 3.10
poetry env use 3.10.6
# Now install the environment
poetry install
# Start the shell
poetry shell
```

Comment (jafermarq) [reply baselines/niid_bench/README.md:80]: We require now all baselines to use python3.10+. A change related to this is also shown in the `pyproject.toml`.

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:84]:
```suggestion
```

Comment (jafermarq) [reply baselines/niid_bench/README.md:84]: I have included these in `pyproject.toml` so they are no longer needed here. Everything is installed via `poetry install` 🪄

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:92]: I haven't looked in absolute detail but at a first glance, it seems all `main_` files are very very similar. Would it be sufficient to have a single `main.py` and then pass a different config depending on what strategy we want to run?
You can specify which config to use with Hydra's `--config-name` argument. Like this:

```
# this runs the content of `main_fedavg.py` but changes the default config (`conf/fedavg_base.yaml`) to `conf/fednova_base`
python -m niid_bench.main_fedavg --config-name fednova_base
```

Comment (jafermarq) [reply baselines/niid_bench/README.md:92]: You know your code better, but if this is possible, it will be ideal to prevent unnecessary code duplication.

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:130]:
```suggestion
| Dataset | partitioning method | FedAvg | SCAFFOLD | FedProx | FedNova |
| :------ | :------ | :---: | :---: | :---: | :---: |
| MNIST | IID<br>Dirichlet (0.5)<br>Sort and Partition (1)<br>Sort and Partition (2)<br>Sort and Partition (3) | | | | |
| FMNIST | IID<br>Dirichlet (0.5)<br>Sort and Partition (1)<br>Sort and Partition (2)<br>Sort and Partition (3) | | | | |
| CIFAR10 | IID<br>Dirichlet (0.5)<br>Sort and Partition (1)<br>Sort and Partition (2)<br>Sort and Partition (3) | | | | |
```

Comment (jafermarq) [reply baselines/niid_bench/README.md:130]: You can use `<br>` to have something similar to latex's `multi-row`. You could use this to make the table in the "Experiment Setup" section to look a bit nicer if you want.

File: baselines/niid_bench/run_all.sh
Comment (jafermarq) [inline baselines/niid_bench/run_all.sh:10]: I'm not sure if it's ideal to run experiment in parallel like this, but if your machine is fine with it, then that's 💯

File: baselines/niid_bench/pyproject.toml
Comment (jafermarq) [inline baselines/niid_bench/pyproject.toml:42]:
```suggestion
python = ">=3.10.0, <3.11.0"  # don't change this
flwr = { extras = ["simulation"], version = "1.5.0" }
```

Comment (jafermarq) [reply baselines/niid_bench/pyproject.toml:42]: Let's use the latest Flower and Python3.10+

File: baselines/niid_bench/pyproject.toml
Comment (jafermarq) [inline baselines/niid_bench/pyproject.toml:44]:
```suggestion
torch = { url = "https://download.pytorch.org/whl/cu113/torch-1.12.1%2Bcu113-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu113/torchvision-0.13.1%2Bcu113-cp310-cp310-linux_x86_64.whl"}
tqdm = "4.66.1"

```

Comment (jafermarq) [reply baselines/niid_bench/pyproject.toml:44]: As mentioned in my comment for the `README.md`, better to install everything in here when possible.

File: baselines/niid_bench/pyproject.toml
Comment (jafermarq) [inline baselines/niid_bench/pyproject.toml:49]:
```suggestion
mypy = "==1.4.1"
```

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:5]:
```suggestion
dataset: [cifar10, mnist, emnist, fashion-mnist]
```

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:20]: and `EMNIST`? (I see it in the code but maybe is not included in the experiments in the end?)

**Review by jafermarq — COMMENTED** [review]
Hi @aashishkolluri,

I have now done an initial pass through your baseline. I was able to follow your instructions and run the code. Great!! You can see below I have left a few comments, most of them very minor, with suggestions to improve different parts of your code (mostly the `readme` and `pyproject.toml` as well as general `main.py`+`conf.yaml` matching)

Please reach out to me if you have some doubts or if you want to discuss the changes I proposed. The majority of them you probably want to apply them directly. Please mark them as resolved when so.

Let me know when your code is ready for a second review. Looking forward to seeing that **MEGA TABLE** at the end of your readme full with numbers!!

File: baselines/niid_bench/run_all.sh
Comment (aashishkolluri) [reply baselines/niid_bench/run_all.sh:10]: I agree and my machine can handle this. I have provided this file as a reference and it can be changed depending on the machine specifications.

File: baselines/niid_bench/run_all.sh
Comment (jafermarq) [inline baselines/niid_bench/run_all.sh:27]: Are these being used in the table you show in the readme? (this comment applies to all `fedprox` experiments where you manually set the proximal term `mu`

File: baselines/niid_bench/niid_bench/conf/fedavg_base.yaml
Comment (jafermarq) [reply baselines/niid_bench/niid_bench/conf/fedavg_base.yaml:15]: why not putting these directly below where the `_target_` is?

File: baselines/niid_bench/niid_bench/conf/fedavg_base.yaml
Comment (jafermarq) [inline baselines/niid_bench/niid_bench/conf/fedavg_base.yaml:40]:
```suggestion
  _target_:  flwr.server.strategy.FedAvg # points to your strategy (either custom or exiting in Flower)
```

File: baselines/niid_bench/niid_bench/conf/fedavg_base.yaml
Comment (jafermarq) [inline baselines/niid_bench/niid_bench/conf/fedavg_base.yaml:34]:
```suggestion
  _target_: niid_bench.models.CNN
```

File: baselines/niid_bench/niid_bench/conf/fedavg_base.yaml
Comment (jafermarq) [reply baselines/niid_bench/niid_bench/conf/fedavg_base.yaml:15]: please do the same with the other configs

**Review by jafermarq — COMMENTED** [review]
Hey @aashishkolluri ,

Thanks for making those fixes. I suggest some other below. Please take a look

Regarding the brief discussion to simplify `run_all.sh`. I'm thinking you can do the following:
* Define a single `main.py` that uses the `fedavg_base.yaml` as default. There is nothing unique for each of the current `mains` so this should be easy to do.
* The difference is that they each load a different `get_client_fn` and `get_evaluate_fn`. You can specify these functions from the configs directly (just like what you do with your strategies) and then call them using `Hydra's` `call` function (instead of `instantiate). 
* I was going to propose a way of using `--multirun` to vastly simplify `run_all.sh` but the explanation was getting to long. I'll propose the changes and you can see if it looks good. I'll do that likely sometime in the weekend. The basic idea is to use the `defaults` keyword in the (new) `base.yaml` so later on  you can do the `--multirun` over each child config file, each to be used for each of your strategies. in that way a single `--multirun` command could be userd, for instance, to generate one row in your table. (but for this i count on you making the changes to use a single main fail)


Ping me when ready.

File: baselines/niid_bench/run_all.sh
Comment (aashishkolluri) [reply baselines/niid_bench/run_all.sh:27]: They are needed for that table. For FedProx I report the best accuracy among these runs for each configuration.

File: baselines/niid_bench/pyproject.toml
Comment (jafermarq) [inline baselines/niid_bench/pyproject.toml:10]: Consider adding your contact info here. Also, add the title of the paper to the `description` field above.

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:100]: Could we add the mean/std results?

**Review by jafermarq — COMMENTED** [review]
Hi @aashishkolluri,

First of all, many thanks for putting the efforts towards making the code much more reusable by getting rid of some duplication. It looks so much better!

I have done a second pass through your baseline, and highlighted some minor things.

Please ping me once you have the completed results. Things look good in general

File: baselines/niid_bench/README.md
Comment (aashishkolluri) [reply baselines/niid_bench/README.md:103]: Added!

File: baselines/niid_bench/niid_bench/main.py
Comment (aashishkolluri) [reply baselines/niid_bench/niid_bench/main.py:36]: Thank you for your suggestions Javier. Perhaps, I will leave it for future work for now.

File: baselines/niid_bench/niid_bench/main.py
Comment (jafermarq) [inline baselines/niid_bench/niid_bench/main.py:109]: Could you add just a couple of lines of code to write a pickle to `save_path` (i.e. the output directory for the experiment) with there content in the `history` object ? Other baselines do this, it is useful to have if people want to create plots later on easily.

File: baselines/niid_bench/pyproject.toml
Comment (jafermarq) [inline baselines/niid_bench/pyproject.toml:87]:
```suggestion

[tool.pylint.typecheck]
generated-members="numpy.*, torch.*, tensorflow.*"

```

Comment (jafermarq) [reply baselines/niid_bench/pyproject.toml:87]: this is needed for the tests. (it was added to the `pyproject.toml` in the template directory after you made your copy.

**Review by jafermarq — COMMENTED** [review]
Hi @aashishkolluri,

Thanks for adding the changes. I have added just a couple of comments. Please take a look.

In addition to those,
* Don't forget you need to run first `./dev/format-baseline.sh <name-of-your-baseline-dir>` and see no errors (it's likely many will be flagged, some automatically corrected). These are easy to fix for the most part.
* After the above, run `./dev/test-baseline.sh <name-of-your-baseline>`. These checks will be run automatically when merging your baseline with `main` but you can run the test on your machine first to check everything is ok.

Please ping me when you have some time to look into these.

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:5]:
```suggestion
dataset: [CIFAR-10, MNIST, Fashion-MNIST]
```

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:11]:
```suggestion
**Paper:** [https://arxiv.org/abs/2102.02079](https://arxiv.org/abs/2102.02079)
```

Comment (jafermarq) [reply baselines/niid_bench/README.md:11]: note how I changed `****` to `**` and also removed the second `:`. This appears in many places in the readme. Could you please fix them all?

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:106]:
```suggestion
| FMNIST | IID<br>Dirichlet (0.5)<br>Sort and Partition (1)<br>Sort and Partition (2)<br>Sort and Partition (3) | 89.23 &#xB1; 0.45<br>88.09 &#xB1; 0.29<br>28.39 &#xB1; 17.09<br>78.10 &#xB1; 2.51<br>82.43 &#xB1; 1.52 | 89.33 &#xB1; 0.27<br>88.44 &#xB1; 0.25<br>10.00 &#xB1; 0.00<br>33.80 &#xB1; 41.22<br>80.32 &#xB1; 5.03 | 89.42 &#xB1; 0.09<br>88.15 &#xB1; 0.42<br>32.65 &#xB1; 6.68<br>78.05 &#xB1; 0.99<br>82.99 &#xB1; 0.48 | 89.36 &#xB1; 0.09<br>88.22 &#xB1; 0.12<br>16.86 &#xB1; 9.30<br>71.67 &#xB1; 2.34<br>81.97 &#xB1; 1.34 |
```

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:98]: I think very few systems can run `--num-processes 7` 😆 ( I was at 100% load using 5 on a 96 thread machine). Should we lower it to 4 here?

File: baselines/niid_bench/README.md
Comment (jafermarq) [inline baselines/niid_bench/README.md:101]: After running your experiments I noticed in several cases (which is expected I guess) not the best performance was achieved in the last round. I assume you reported the best observed result (then averaged those over the several seeds you used). I think it's worth indicating this (in case people only look into the last round acc)

**Review by jafermarq — COMMENTED** [review]
Hi @aashishkolluri ,

Thanks for including the changes. I have highlighted a few minor ones. I also run all the experiments. It looks good. 

But it seems the `./dev/test-baseline.sh ...` command still shows quite a few errors. For those in the form `too-many-` you can disable them by placing a line like `#pylint: disable=too-many-<...>` next or above the line triggering the error. But please don't use the `disable` option for other errors. If there is some error you are having a hard time to fix, let me know and I can probably help.

Once that test passes, we can merge the baseline.

**Review by jafermarq — COMMENTED** [review]
Hi @aashishkolluri ,

Thanks for applying those many changes since my last review. Your code and results look good so i'm happy to merge this baseline. But before we do this just one small thing: part of the tests check for the existence of `client.py`, `server.py`, `strategy.py` (which you didn't have). I added them as pretty much empty files. But i used that opportunity to add your custom strategies into `strategy.py` and I made the necessary changes elsewhere in the code (including the configs) so everything should work as before. 

➡️ Could you please double check the changes i did in my last commit?

Also, if there is anything you'd like to add or modify to your readme, now it's the time.

Let me know when you have a couple of minutes to double check the above. Maybe running a couple of experiments to see everything remains the same won't hurt.

All test in the GitHub CI pass

**Comment by aashishkolluri** [comment]
> Hi @aashishkolluri ,
> 
> Thanks for applying those many changes since my last review. Your code and results look good so i'm happy to merge this baseline. But before we do this just one small thing: part of the tests check for the existence of `client.py`, `server.py`, `strategy.py` (which you didn't have). I added them as pretty much empty files. But i used that opportunity to add your custom strategies into `strategy.py` and I made the necessary changes elsewhere in the code (including the configs) so everything should work as before.
> 
> ➡️ Could you please double check the changes i did in my last commit?
> 
> Also, if there is anything you'd like to add or modify to your readme, now it's the time.
> 
> Let me know when you have a couple of minutes to double check the above. Maybe running a couple of experiments to see everything remains the same won't hurt.
> 
> All test in the GitHub CI pass

Hi @jafermarq ,

I have checked your changes in the last commit and everything seems fine to me. I have also run a few commands from the Readme again to check the results and they are as expected.

**Review by jafermarq — APPROVED** [review]
@aashishkolluri , thanks for confirming. This baseline is ready to merge then! amazing work 🚀

**Review by danieljanes — APPROVED** [review]

**Pull request merged**
